saleor/graphql/app/dataloaders.py

```diff
@@ -35,3 +38,49 @@ def batch_load(self, keys):
             extensions_map[extension.app_id].append(extension)
             app_extension_loader.prime(extension.id, extension)
         return [extensions_map.get(app_id, []) for app_id in keys]
+
+
+def get_app(raw_auth_token) -> Union[None, App]:
+    if raw_auth_token is None:
+        return None
+    tokens = AppToken.objects.filter(token_last_4=raw_auth_token[-4:]).values_list(
+        "app_id", "auth_token"
+    )
+    app_ids = [
+        app_id
+        for app_id, auth_token in tokens
+        if check_password(raw_auth_token, auth_token)
+    ]
+    return App.objects.filter(id__in=app_ids, is_active=True).first()
+
+
+class AppByTokenLoader(DataLoader):
+    context_key = "app_by_token"
+
+    def batch_load(self, keys):
+        last_4s_to_raw_token_map = defaultdict(list)
+        for raw_token in keys:
+            last_4s_to_raw_token_map[raw_token[-4:]].append(raw_token)
+
+        tokens = (
+            AppToken.objects.using(self.database_connection_name)
+            .filter(token_last_4__in=last_4s_to_raw_token_map.keys())
+            .values_list("auth_token", "token_last_4", "app_id")
+        )
+        authed_apps = {}
+        for auth_token, token_last_4, app_id in tokens:
+            for raw_token in last_4s_to_raw_token_map[token_last_4]:
+                if check_password(raw_token, auth_token):
+                    authed_apps[raw_token] = app_id
+
+        apps = App.objects.using(self.database_connection_name).filter(
+            id__in=authed_apps.values(), is_active=True
+        )
+
+        return [apps.filter(id=authed_apps.get(key)).first() for key in keys]
+
+
+def load_app(context):
+    if not getattr(context, "auth_token", None):
+        return None
+    return AppByTokenLoader(context).load(context.auth_token).get()
```

Comment:
How does it works exactly? From the usages I see we don't have to "wait" or resolve a promise with "then", so does the `get` call resolve the app immediately? If so, is `load_app` ensuring the constant number of DB queries, regardless of how many times it was called during a single request?

Reply:
It works very alike to a LazyObject but is more eager - it fetches the App from DB asap and caches it for future. Yes, right now this increases the query count in some places. I'm in the process of teaching Django to understand Promises and after that the `.get()` could go someplace else to work more lazily.